File: clojure/hamming/src/hamming.clj
(ns hamming)

(defn distance
  "Returns the hamming distance between two DNA strands."
  [strand1 strand2]
  (when (not= (count strand1) (count strand2)) 
    (throw (IllegalArgumentException. "strands must be of equal length")))
  (count (filter identity (map not= strand1 strand2))))

File: elixir/hamming/lib/hamming.ex
defmodule Hamming do
  @doc """
  Returns number of differences between two strands of DNA, known as the Hamming Distance.

  ## Examples

  iex> Hamming.hamming_distance(~c"AAGTCATA", ~c"TAGCGATC")
  {:ok, 4}
  """
  @spec hamming_distance([char], [char]) :: {:ok, non_neg_integer} | {:error, String.t()}
  def hamming_distance(strand1, strand2) when length(strand1) != length(strand2), do:
    {:error, "strands must be of equal length"}
  def hamming_distance(strand1, strand2), do:
    {
      :ok, 
      Enum.zip_with(strand1, strand2, &Kernel.!=/2)
      |> Enum.filter(&Function.identity/1)
      |> Enum.count()
    }
end
